tree/ntuple/v7/src/RClusterPool.cxx

```diff
@@ -120,13 +121,64 @@ void ROOT::Experimental::Detail::RClusterPool::ExecReadClusters()
          }
       }
 
+      // At this point, but not before, we can create an RPageSink with the same metadata held by fPageSource
+      // Before this point, fPageSource still doesn't have all the header metadata so it's pointless.
+      if (!fPageSink){
+
+         // HARDCODED FILENAME FOR THE CACHED RNTUPLE
+         std::string_view cachedntuplepath = "cachedntuple.root";
+
+         fPageSink = RPageSink::Create(fPageSource.GetNTupleName(), cachedntuplepath);
+         auto modelptr = fPageSource.GetDescriptor().GenerateModel()->Clone();
+         fPageSink->Create(*modelptr);
+      }
+
+      // Track the number of entries seen so far
+      // Needed in the call to fPageSink->CommitCluster that will be done in CacheCluster
+      ClusterSize_t entriessofar{0};
       for (auto &item : readItems) {
-         if (item.fClusterId == kInvalidDescriptorId)
+         if (item.fClusterId == kInvalidDescriptorId){
+            // Need to commit the cached dataset if present
+            fPageSink->CommitDataset();
             return;
+         }
 
          // TODO(jblomer): the page source needs to be capable of loading multiple clusters in one go
          auto cluster = fPageSource.LoadCluster(item.fClusterId, item.fColumns);
 
+         // Cache the current cluster
+         auto CacheCluster = [&](DescriptorId_t clusterId){
+            const auto &clusterDesc = fPageSource.GetDescriptor().GetClusterDescriptor(clusterId);
+            const auto clusterentries = clusterDesc.GetNEntries();
+
+            // Traverse columns in cluster
+            for (auto columnId: clusterDesc.GetColumnIds()){
+
+               const auto &pageRange = clusterDesc.GetPageRange(columnId);
+               std::uint32_t firstElementInPage = 0;
+
+
+               RPageStorage::RSealedPage sealedPage;
+               // Traverse pages in column
+               for (const auto &pi : pageRange.fPageInfos) {
+
+                  auto buffer = std::make_unique<unsigned char []>(pi.fLocator.fBytesOnStorage);
+                  sealedPage.fBuffer = buffer.get();
+                  fPageSource.LoadSealedPage(columnId, RClusterIndex(clusterId, firstElementInPage), sealedPage);
```

Comment:
The `LoadCluster()` call above will read all the required pages.  Wouldn't this trigger an additional read request?

Reply:
Not sure how to do it in another way for now. If I entered the above call to `LoadCluster` I would also need to optionally pass some activator for the caching logic. I need to think about this a bit more